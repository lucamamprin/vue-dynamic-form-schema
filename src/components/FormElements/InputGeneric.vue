<template>
  <div
    class="input floating-label"
    :class="[
      {
        'input-transparent': darkMode,
        'mt-0': resetMarginTop,
        'floating-label-floated': value || floatedLabel,
      },
      $attrs.type,
    ]"
  >
    <div class="position-relative">
      <input
        v-bind="$attrs"
        :class="{
          'border-danger': hasError,
        }"
        :aria-describedby="`${describedBy} ${maxLengthDescribedBy}`"
        :value="value"
        :name="name"
        :id="id"
        :disabled="disabled"
        :min="min"
        :max="max"
        :minlength="minLength"
        :maxlength="maxLength"
        :placeholder="placeholderText"
        :aria-labelledby="`label-${name}`"
        :aria-invalid="invalid.toString()"
        :aria-required="required.toString()"
        v-on="{
          ...$listeners,
          input: (event) =>
            $emit(
              'input',
              uppercase ? event.target.value.toUpperCase() : event.target.value
            ),
        }"
      >
      <label
        :id="`label-${name}`"
        :for="id"
        v-html="label"
      />
    </div>

    <div
      v-if="hasCharacterCount || hasLeftColSlot"
      class="pal-row pal-no-gutters mt-2"
    >
      <div
        v-if="hasLeftColSlot"
        class="pal-col px-0"
      >
        <slot name="leftCol" />
      </div>

      <div
        v-if="hasCharacterCount"
        class="pal-col px-0"
      >
        <max-length
          :typed-text="value"
          :sr-character-count="maxLengthDescribedBy"
          :label="removeStar(label)"
          :max-length="maxLength"
        />
      </div>
    </div>

    <div v-if="hasError && errorMessage">
      <error-message
        :id="describedBy"
        :error-message="errorMessage"
      />
    </div>
  </div>
</template>

<script>
import removeStar from "../../utilities/removeStar"
const MaxLength = () => import("./Atoms/MaxLength")
const ErrorMessage = () => import("./Atoms/ErrorMessage")

export default {
  name: "InputGeneric",
  inheritAttrs: false,
  mixins: [
    removeStar,
  ],
  components: {
    MaxLength,
    ErrorMessage,
  },
  computed: {
    hasLeftColSlot () {
      return this.$slots.leftCol
    },
    hasCharacterCount () {
      return this.maxLength && this.showCharacterCount
    },
    maxLengthDescribedBy () {
      return this.hasCharacterCount ? `${this.id}-counter` : ""
    },
  },
  props: {
    floatedLabel: {
      type: Boolean,
      default: false,
    },
    id: {
      type: String,
      required: true,
    },
    describedBy: {
      type: String,
      required: true,
    },
    value: {
      type: [
        String, Number,
      ],
      default: null,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      required: true,
    },
    errorMessage: {
      type: String,
      default: null,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    min: {
      type: null,
      default: null,
    },
    max: {
      type: null,
      default: null,
    },
    minLength: {
      type: Number,
      default: null,
    },
    maxLength: {
      type: Number,
      default: null,
    },
    showCharacterCount: {
      type: Boolean,
      default: true,
    },
    resetMarginTop: {
      type: Boolean,
      default: false,
    },
    darkMode: {
      type: Boolean,
      default: false,
    },
    placeholderText: {
      type: String,
      default: " ",
    },
    invalid: {
      type: Boolean,
    },
    required: {
      type: Boolean,
      default: false,
    },
    uppercase: {
      type: Boolean,
      default: false,
    },
  },
}
</script>
