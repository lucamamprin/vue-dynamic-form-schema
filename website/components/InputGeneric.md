# InputGeneric

## Props

<!-- @vuese:InputGeneric:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|value|-|`String` /  `Number`|`false`|null|
|type|-|`String`|`false`|text|
|min|-|—|`false`|null|
|max|-|—|`false`|null|

<!-- @vuese:InputGeneric:props:end -->


## Slots

<!-- @vuese:InputGeneric:slots:start -->
|Name|Description|Default Slot Content|
|---|---|---|
|leftCol|-|-|

<!-- @vuese:InputGeneric:slots:end -->


## MixIns

<!-- @vuese:InputGeneric:mixIns:start -->
|MixIn|
|---|
|props|
|textInputs|

<!-- @vuese:InputGeneric:mixIns:end -->


