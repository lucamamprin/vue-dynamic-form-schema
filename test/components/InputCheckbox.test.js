import Vue from "vue"
import { shallowMount } from "@vue/test-utils"
import InputGeneric from "../../src/components/FormElements/InputGeneric.vue"

const props = {
  label: "Check me out",
  name: "checkme",
  id: "check",
  describedBy: "checkme-describe",
}

let wrapper

describe("InputCheckbox", () => {
  beforeEach(() => {
    wrapper = shallowMount(InputGeneric, {
      propsData: props,
    })
  })

  test("mounts correctly", () => {
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
