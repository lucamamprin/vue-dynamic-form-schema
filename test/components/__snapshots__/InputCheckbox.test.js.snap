// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputCheckbox mounts correctly 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input aria-describedby="checkme-describe " name="checkme" id="check" placeholder=" " aria-labelledby="label-checkme" aria-invalid="false" aria-required="false" class=""> <label id="label-checkme" for="check">Check me out</label></div>
  <!---->
  <!---->
</div>
`;
